topics/w5-programming-by-demonstration.md

```diff
@@ -9,3 +9,25 @@
+In terms of learning though, just like Flash Fill, it does generalize on the performed actions, and learns the patterns. 
+
+
+The limits of programming by demonstration haven't been pushed, they aren't even in sight yet. With the recent development boom in the area of artificial intelligence, PbD can defintely benefit from it. Having a houserobot do the chores doesn't seem as far now that many companies have developed robots that already are doing regular tasks, like barista robots and warehouse robots. To make one of these robots perform a new chore, you could, for example, demonstrate the desired chore, and the robot will then be able to replicate that chore. Another avenue can be a creation of an application, where a designer can show prototyped user interfaces and how they should transition, and the PbD program could program the app to look and perform like desired. The future for programming by demonstration is very promising, since the result could be programming of actions, without the need for a programmer. At least this is the dream of many who are not versed in programming, since programming development can take quite a long time, and relies on the communication with the client, on what the client wants in the product. The premise of reaching a point, where no programmer is needed, is that the program has to understand what the client wants, better then a group of trained programmers who are also human. Additionally the assumption is that the client knows what they want and how it should be implemented. Even with the development of ai, we have not reached a point, where the ai can program better then the professionals. What we have reached is that the ai can help professionals write code faster. Combine the ai with task specific heuristics and a professional, and you get that PbD will help developers, like many other tools, develop faster and there lies the true opportunity for programming by demonstration algorithms.
```

Comment:
Would be good to reflect on the general challenges of PBDs.
- Accuracy: demonstrations can be ambiguous or incomplete
- Trust and explainability: how PBD systems communicate what the system has learned in an understandable way
- Context awareness: PBD systems need to maintain an understanding of the context across multiple demonstrations
- Other challenges